### DISCUSSION TITLE
Proposal to use pp-based scoring

### DISCUSSION THREAD

**Discussion by Pasi4K5** [ppy/osu#24159] [Ideas] [answered]
Since there's currently a discussion about scoring, I'd like to share this idea that came to me since I think it solves a lot of problems and just makes sense.

The way I see it, there's 3 main problems we have to deal with:
1. The order of a leaderboard might differ depending on whether it is sorted by score or sorted by pp.
2. There will always be discussions on things like accuracy/combo-spilt, so locking in a scoring system now and never change it is kind of dangerous.
3. Scoring should be at least mostly skill-based.

I think a good way to address this is to not have a scoring algorithm at all, but use pp instead, while still keeping the concept of standardised scoring (Meaning 1.000.000 score = nomod SS). This can be achieved by making the score proportional to the pp achieved, with 1.000.000 score being equivalent to the maximum pp achievable without any mods.

Here's an example:

Let's say there's a map on which a nomod SS is worth 100pp. If you set a 100% score with nomod, you get 1.000.000 score as usual. If, let's say, there's a few misses in your play and you only get 75pp, your score will be 750.000. And if you play with DT and you set a 180pp play, your score will be 1.800.000.

I think this would solve a lot of problems:
1. The pp system is designed to be skill-based and it's the closest thing we have to objectively judge how good a score is, so why wouldn't we just use it for scoring?
2. There's no need to add the option to sort leaderboards by pp, since the order would be the same anyway.
3. There's no need to lock in a scoring system now. The pp algorithm always changes, and I don't see why scoring can't do this in the same way.
4. It just makes sense. Especially for someone who is new to the game, having 2 kinds of "scoring" (score and pp) doesn't make much sense if they differ from each other.
5. There's no need for arbitrary mod multipliers, and problems like mod multipliers in combination with mod customisation (like "Speed increase" for DT) would just vanish, because pp can handle that.

I am aware of this point that @peppy brought up in the recent community meeting:
"You expect a scoring algorithm to be defined by the basic stats, not by intrinsics of a map."
But, the way I see it, neither ScoreV1 nor ScoreV2 is simple enough for this to matter. If the scoring system was as basic as "all your 50s, 100s and 300s get summed up, and that's your score", then I would get this point. But because the scoring system is too complex for most players to easily understand or care about anyway, I think we might as well use the more skill-based algorithm that's already established and has all the above benefits.

Another argument against this (this was brought up by BTMC in the same meeting):
"The scoring system should be simple enough where if a player is new to the game, it should be very easy to explain."
But is that really the case with ScoreV2? I don't think so. I've been playing osu! for a few years now, and I for my part don't know exactly how scoring works. If I were to explain ScoreV2 to a new player, I would just say "A perfect play without mods is 1.000.000 points. If you do better/worse or play with mods that increase/decrease the difficulty, you'll earn more/less score." If we would use this suggested system, you can still explain it in the exact same way. If a new player asks me how the scoring works, I'm not gonna give them a formula and tell them about the math behind it. The exact math behind the scoring algorithm is too complex (and irrelevant, for that matter) for most players to care anyway. The only thing you need to know as a player is: you do better -> you get more score. Additionally, having 2 different types of "scoring" (score and pp) would most likely add confusion for the new player, which is why IMO this is not really a valid argument.

Those are the 2 arguments I've heard against this idea so far, but since (to my knowledge) this idea hasn't really been brought up or discussed properly yet, I'm sure there will be more arguments against this that I haven't thought about, and I'd be more than happy to discuss them.

**Comment by bdach** [comment]
Few extra objective counterarguments (that aren't "I don't like the idea" - well, I don't, but that can't be reasoned with):

- The pp algorithm is massively skewed towards the top end. Beginner players will routinely get pp values in the ones. How would that scale so that it doesn't seem massively unsatisfying to new players?
- pp-based scoring implies the maximum number of points that can be gained on a map can change depending on map, unless score is renormalised to the max number of pp achievable on the map.
- At some point the goal was to add a "score" breakdown on the results screen, that would indicate how much of the final score was gained from accuracy and how much was gained from combo. This would be infeasible if pp was to be what score is derived from. Read https://github.com/ppy/osu/pull/16483#discussion_r787173741 for more.

**Reply by Pasi4K5** [reply]
I think you misunderstood my idea. The maximum (nomod) score on a map stays at 1.000.000, it just scales with pp.

aka:
`score = 1_000_000 * achieved_pp / max_pp`
(where `max_pp` is the pp achieved when SSing the map with nomod)

**Reply by bdach** [reply]
>  unless score is renormalised to the max number of pp achievable on the map

Doesn't change anything about either of the remaining points.

**Reply by Pasi4K5** [reply]
> The pp algorithm is massively skewed towards the top end. Beginner players will routinely get pp values in the ones. How would that scale so that it doesn't seem massively unsatisfying to new players?

If it is renormalised, then this wouldn't be a problem. New players would still get scores in the hundred thousands if they play maps that fit their skill level.

> At some point the goal was to add a "score" breakdown on the results screen, that would indicate how much of the final score was gained from accuracy and how much was gained from combo. This would be infeasible if pp was to be what score is derived from. Read https://github.com/ppy/osu/pull/16483#discussion_r787173741 for more.

Frankly, I'm not exactly familiar with pp calculation, but afaik pp is calculated by "grading" different skills (accuracy, speed, ect.), so maybe it could be possible to break it down that way?

**Reply by bdach** [reply]
> If it is renormalised, then this wouldn't be a problem. New players would still get scores in the hundred thousands if they play maps that fit their skill level.

No, I believe you'd find it still very much would. Take a map with a 50pp perfect play and give it to a beginner who can get 2-3pp out of it at most. That means that the player's effective score cap is 40-60k. Out of 1m.

> Frankly, I'm not exactly familiar with pp calculation, but afaik pp is calculated by "grading" different skills (accuracy, speed, ect.), so maybe it could be possible to break it down that way?

Please read the linked comment, then the pp code, then come back to me with a proposal. Skills are a concept in pp calculation but the final pp value is not a trivial sum of them or anything. The skill values are combined to the final pp value in a way that is not linearly separable.

**Comment by WitherFlower** [comment]
Ordering map leaderboards by performance points is not that bad of an idea, until you consider the fact that pp is, and will probably forever be a system in development. 

Let me just ask this : do you want to see every leaderboard on every map being shuffled at least once every year ?

I know I don't, and I doubt you can make everyone in the community agree that this is a good direction.

**Reply by bdach** [reply, marked as answer]
I somehow forgot about this, and this is indeed the reaper for the entire concept, even speaking from the infrastructure side of things. It would mean that _every single user's score totals_ would have to be recomputed on _every pp rebalance_. Which is basically infeasible.

**Reply by Pasi4K5** [reply]
> I somehow forgot about this

Aaaand so did I. Well, I guess it would be possible to just calculate the score once and then *don't* recalculate it after a pp rebalance, but that defeats the whole point to a certain degree, so I think at this point we can discard this idea.

**Reply by Rekunan** [reply]
The better solution anyway is to add pp as a separate sort method, meaning only one sort would have to update instead of all.

**Comment by Zyfarok** [comment]
Three more issues with pp as a score that are not mentioned here :
- pp is determined ONLY on the data that is stored in the database, that is : currently only max_combo and amount of 300/100/50/miss/etc... If no data is stored regarding non-max combos, then no matter the pp algorithm, it will have to take a (safe / lower-estimate) guess. Also, if it is decided that future plays have more data stored to make a more accurate pp prediction, previous scores will not be able to benefit from that. (By the way, if a good scoring algorithm is made, it could even be used as input of the pp algorithm...)
- pp is meant to be a "safe underestimate", that is, no map/score should give much more than other pp maps/scores, because otherwise it would give an unfair advantage to people playing/making that one map/score, but it's ok to have a map/score give much less than what it should give, because the player can just pp with other maps/scores. This is for example why non-FCs are usually giving much less pp than they should give, just to be "safe", which is unacceptable for scoring.
- the pp formula is complex and player don't understand as easily where pp comes from compared to score, and when watching a 1v1 for example, it would not necessarily be easy to predict who is gonna win more score on the next note. On the other hand, you know that with a good scoring formulas, the guy with higher combo and accuracy is guaranteed to gain more points on the next note if he taps it right on time.

TL;DR : A good scoring formula doesn't need to account for the actual difficulty of a map, but it needs to account for every notes and combo in the map, it shouldn't be an "approximation" (unless we can't do better, like when converting old plays), and it should be simple enough such that experienced players/viewers/commentators can understand the basic trends of how combo/acc/... affects it and how it evolves during the play.

**Reply by Rekunan** [reply]
I want to make some minor corrections here:

- The second half can be similarly used with score, only x amount of data is currently used for it, and if more data is stored for future plays, past scores will not be able to benefit from that.

- To be more specific, pp assumes the worst case scenario for everything (misses, acc drops, etc.).

- While pp formula is complex, players understand that harder sections, in terms of mechanical difficulty, have the potential to award more pp, with the actual award being dependent on how well they play the sections; and it's actually been acknowledged before that players understand pp more than score (I assume cause of scorev1ing and similar).
